src/main/java/gregtech/common/metatileentities/multi/electric/MetaTileEntityAssemblyLine.java

```diff
@@ -60,4 +70,127 @@ public ICubeRenderer getBaseTexture(IMultiblockPart sourcePart) {
     }
+
+    private int beamCount;
+    @SideOnly(Side.CLIENT)
+    private GTLaserBeamParticle[][] beamParticles;
+
+    @Override
+    public void update() {
+        super.update();
+        if(ConfigHolder.client.shader.assemblyLineParticles) {
+            if (getRecipeMapWorkable().isWorking() || getRecipeMapWorkable().isActive()) {
+                int maxBeams = getAbilities(MultiblockAbility.IMPORT_ITEMS).size() + 1;
+                int maxProgress = getRecipeMapWorkable().getMaxProgress();
+
+                // Each beam should be visible for an equal amount of time, which is derived from the maximum number of
+                // beams and the maximum progress in the recipe.
+                int beamTime = Math.max(1, maxProgress / maxBeams);
+
+                int currentBeamCount = Math.min(maxBeams, getRecipeMapWorkable().getProgress() / beamTime);
+
+                if (currentBeamCount != beamCount) {
+                    beamCount = currentBeamCount;
+                    writeCustomData(GregtechDataCodes.UPDATE_PARTICLE, this::writeParticles);
+                }
+                return;
+            }
+            if (beamCount != 0) {
+                beamCount = 0;
+                writeCustomData(GregtechDataCodes.UPDATE_PARTICLE, this::writeParticles);
+            }
+        }
+    }
+
+    @Override
+    public void writeInitialSyncData(PacketBuffer buf) {
+        super.writeInitialSyncData(buf);
+        writeParticles(buf);
+    }
+
+    @Override
+    public void receiveInitialSyncData(PacketBuffer buf) {
+        super.receiveInitialSyncData(buf);
+        readParticles(buf);
+    }
+
+    @Override
+    public void receiveCustomData(int dataId, PacketBuffer buf) {
+        if (dataId == GregtechDataCodes.UPDATE_PARTICLE) {
+            readParticles(buf);
+        } else {
+            super.receiveCustomData(dataId, buf);
+        }
+    }
+
+    @Override
+    public void onRemoval() {
+        super.onRemoval();
+        if (getWorld().isRemote && beamParticles != null) {
+            for (GTLaserBeamParticle[] particle : beamParticles) {
+                if (particle[0] != null) {
+                    particle[0].setExpired();
+                    particle[1].setExpired();
+                }
+            }
+        }
+        beamParticles = null;
+    }
+
+    private void writeParticles(PacketBuffer buf) {
+        buf.writeVarInt(beamCount);
+    }
+
+    private void readParticles(PacketBuffer buf) {
+        beamCount = buf.readVarInt();
+        if (beamParticles == null) {
+            beamParticles = new GTLaserBeamParticle[17][2];
+        }
+        for (int i = 0; i < beamParticles.length; i++) {
+            GTLaserBeamParticle particle = beamParticles[i][0];
+            if (i < beamCount && particle == null) {
+                Vector3 startPos = new Vector3().add(
+                                getPos().offset(getFrontFacing().rotateY().getOpposite(), i))
```

Comment:
I would suggest using a single MutableBlockPos here, and using it to set the position for each iteration, in order to reduce object creation.